src/main/java/seedu/address/logic/commands/RemoveTagCommand.java

```diff
@@ -49,19 +60,61 @@ public CommandResult executeUndoableCommand() throws CommandException {
         requireNonNull(model);
         try {
             if (index.orElse(null) == null) {
-                model.removeTag(toRemove);
-                return new CommandResult(String.format(MESSAGE_SUCCESS));
+                removeAllTag(toRemove);
             } else {
-                model.removeTag(index.get(), toRemove);
-                return new CommandResult(String.format(MESSAGE_SUCCESS));
+                remvoveOneTag(index.get(), toRemove);
             }
-        } catch (DuplicatePersonException e) {
+            return new CommandResult(String.format(MESSAGE_SUCCESS));
+        } catch (DuplicatePersonException dpe) {
             throw new CommandException(MESSAGE_DUPLICATE_PERSON);
-        } catch (PersonNotFoundException e) {
+        } catch (PersonNotFoundException pnfe) {
             throw new CommandException(MESSAGE_NOT_DELETED);
         }
     }
 
+    /**
+     * Ensures that the index given can be found within the given list.
+     * @param index of the person in the filtered person list to remove tag from.
+     * @throws CommandException if the index given is out of the bounds of the filtered list.
+     */
+    private void requireIndexValid(Index index) throws CommandException {
+        List<ReadOnlyPerson> lastShownList = model.getFilteredPersonList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        }
+    }
+
+    /**
+     * Removes a tag from a specified person.
+     * @param index of the person in the filtered person list to remove tag from.
+     * @param toRemove is the tag to be removed from the person
+     * @throws CommandException if the tag to be removed does not exist in the person.
+     */
+    private void remvoveOneTag(Index index, Tag toRemove) throws PersonNotFoundException, DuplicatePersonException,
+            CommandException {
+        requireIndexValid(index);
+        try {
+            model.removeTag(index, toRemove);
+        } catch (NoSuchTagException nste) {
+            throw new CommandException(String.format(MESSAGE_TAG_NOT_FOUND_IN, index.getOneBased()));
+        }
+    }
+
+    /**
+     * Removes a tag from the whole addressbook.
+     * @param toRemove is the tag to be removed from the addressbook.
+     * @throws CommandException if tag does not exist in the addressbook.
+     */
+    private void removeAllTag(Tag toRemove) throws PersonNotFoundException, DuplicatePersonException,
```

Comment:
Mind elaborating, when removing the tag from everybody, how will `PersonNotFoundException` and `DuplicatePersonException` be thrown?